provider/oci/environ.go

```diff
@@ -123,54 +347,344 @@ func (e *Environ) SetConfig(cfg *config.Config) error {
+	// var types []instances.InstanceType
+
+	if args.ControllerUUID == "" {
+		return nil, errors.NotFoundf("Controller UUID")
+	}
+
+	networks, err := e.ensureNetworksAndSubnets(ctx, args.ControllerUUID, e.Config().UUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	zones, err := e.AvailabilityZones(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	zone := zones[0].Name()
+	network := networks[zone][0]
+	// refresh the global image cache
+	// this only hits the API every 30 minutes, otherwise just retrieves
+	// from cache
+	imgCache, err := refreshImageCache(e.Compute, e.ecfg().compartmentID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Debugf("Image cache contains: %v", imgCache)
+	series := args.Tools.OneSeries()
+	arches := args.Tools.Arches()
+
+	types := imgCache.SupportedShapes(series)
+
+	defaultType := string(VirtualMachine)
+	if args.Constraints.VirtType == nil {
+		args.Constraints.VirtType = &defaultType
+	}
+
+	// check if we find an image that is compliant with the
+	// constraints provided in the oracle cloud account
+	args.ImageMetadata = imgCache.ImageMetadata(series, *args.Constraints.VirtType)
+
+	spec, image, err := findInstanceSpec(
+		args.ImageMetadata,
+		types,
+		&instances.InstanceConstraint{
+			Series:      series,
+			Arches:      arches,
+			Constraints: args.Constraints,
+		},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	tools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("agent binaries: %v", tools)
+	if err = args.InstanceConfig.SetTools(tools); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if err = instancecfg.FinishInstanceConfig(
+		args.InstanceConfig,
+		e.Config(),
+	); err != nil {
+		return nil, errors.Trace(err)
+	}
+	hostname, err := e.namespace.Hostname(args.InstanceConfig.MachineId)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	tags := args.InstanceConfig.Tags
+
+	var apiPort int
+	var desiredStatus ociCore.InstanceLifecycleStateEnum
+	// Wait for controller to actually be running
+	if args.InstanceConfig.Controller != nil {
+		apiPort = args.InstanceConfig.Controller.Config.APIPort()
+		desiredStatus = ociCore.InstanceLifecycleStateRunning
+	} else {
+		desiredStatus = ociCore.InstanceLifecycleStateProvisioning
+	}
+
+	cloudcfg, err := e.getCloudInitConfig(series, apiPort)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
+
+	// compose userdata with the cloud config template
+	logger.Debugf("Composing userdata")
+	userData, err := providerinit.ComposeUserData(
+		args.InstanceConfig,
+		cloudcfg,
+		OCIRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	var rootDiskSize int
+	if args.Constraints.RootDisk != nil {
+		rootDiskSize = int(*args.Constraints.RootDisk)
+		if int(*args.Constraints.RootDisk) < MinVolumeSizeMB {
```

Comment:
Should we:
a) log a message with the actual requested size or;
b) return an error an error indicating the operator-supplied constraints could not be satisfied?

Reply:
I would at most log it. An error seems too extreme. I think it's much more acceptable to set minimum or maximum sizes if any of the limits are not satisfied.

Reply:
I think logging is reasonable. Letting the user understand why you aren't honoring their request is a good thing.

It is probably a Warning level, because we are explicitly ignoring the user value.